File: JFFliter/JFFliter/components/JFFliterSectionPickerCell.m
//
//  JFFliterSectionPickerCell.m
//  JFTX_WuHanPJJY
//
//  Created by geeksprite on 2017/7/25.
//  Copyright © 2017年 jf. All rights reserved.
//

#import <Masonry.h>
#import "JFFliterItem.h"
#import "JFAreaPickerView.h"
#import "JFSelectDatePickView.h"
#import "JFFliterAppearanceManager.h"
#import "JFFliterSectionPickerCell.h"
#import "JFVarificationInputManager.h"
#import "UITextField+CustomPlacehold.h"

@interface JFFliterSectionPickerCell ()<UITextFieldDelegate>

@property (nonatomic, strong) UIView *pickerContainerView;
@property (nonatomic, strong) UIView *indicateView;
@property (nonatomic, strong) UITextField *leftTF;
@property (nonatomic, strong) UITextField *rightTF;
@property (nonatomic, strong) UITextField *singleTF;
@property (nonatomic, strong) JFSelectDatePickView *leftPickerView;
@property (nonatomic, strong) JFSelectDatePickView *rightPickerView;
@property (nonatomic, strong) JFAreaPickerView *leftAddressPicker;
@property (nonatomic, strong) JFAreaPickerView *rightAddressPicker;

@end

@implementation JFFliterSectionPickerCell

#pragma amrk - Override
- (void)setCellItem:(JFFliterItem *)cellItem {
    [super setCellItem:cellItem];

    self.indicateView.hidden = !self.cellItem.sectionSelect;
    self.rightTF.hidden = !self.cellItem.sectionSelect;
    [self.leftTF customTextfiledPlaceholdWithColor:[JFFliterAppearanceManager sharedManager].pickerPlaceHolderColor content:self.cellItem.leftPlaceHolder ? self.cellItem.leftPlaceHolder : @"" fontSize:[FontUtil fontSize:14]];
    [self.rightTF customTextfiledPlaceholdWithColor:[JFFliterAppearanceManager sharedManager].pickerPlaceHolderColor content:self.cellItem.rightPlaceHolder ? self.cellItem.rightPlaceHolder : @"" fontSize:[FontUtil fontSize:14]];
    switch (cellItem.pickerType) {
        case JFFliterPickerTypeTextField:
        {
            self.leftTF.inputView = nil;
            self.rightTF.inputView = nil;
        }
            break;
        case JFFliterPickerTypeDatePicker:
        {
            @weakify(self);
            [self workWhenRunloopIdle:^{
                @strongify(self);
                self.leftPickerView.myPickerView.maximumDate = self.cellItem.leftMaxDate;
                self.leftPickerView.myPickerView.minimumDate = self.cellItem.leftMinDate;
                self.rightPickerView.myPickerView.maximumDate = self.cellItem.rightMaxDate;
                self.rightPickerView.myPickerView.minimumDate = self.cellItem.rightMinDate;
                self.leftTF.inputView = self.leftPickerView;
                self.rightTF.inputView = self.rightPickerView;
                self.leftPickerView.block = [self pickerBlockForTextField:self.leftTF];
                self.rightPickerView.block = [self pickerBlockForTextField:self.rightTF];
            }];
        }
            break;
        case JFFliterPickerTypeDatePickerAddress:
        {
            @weakify(self);
            [self workWhenRunloopIdle:^{
                @strongify(self);
                self.leftTF.inputView = self.leftAddressPicker;
                self.rightTF.inputView = self.rightAddressPicker;
                self.leftAddressPicker.tfBlock = [self addressPickerBlockForTextField:self.leftTF];
                self.rightAddressPicker.tfBlock = [self addressPickerBlockForTextField:self.rightTF];
            }];
        }
            break;
        default:
            break;
    }
    if (self.cellItem.shouldReset) {
        self.cellItem.shouldReset = NO;
        self.leftTF.text = nil;
        self.rightTF.text = nil;
    }
}

- (void)didInit {
    [super didInit];
    [self addSubview:self.pickerContainerView];
    [self.pickerContainerView addSubview:self.leftTF];
    [self.pickerContainerView addSubview:self.indicateView];
    [self.pickerContainerView addSubview:self.rightTF];
}

- (void)didLayoutSubviews {
    [super didLayoutSubviews];
    UIEdgeInsets insets = UIEdgeInsetsMake(0, [JFFliterAppearanceManager sharedManager].sectionPaddingSize.width, [JFFliterAppearanceManager sharedManager].sectionPaddingSize.height, [JFFliterAppearanceManager sharedManager].sectionPaddingSize.width);
    
    [self.pickerContainerView mas_makeConstraints:^(MASConstraintMaker *make) {
        make.edges.equalTo(self).with.insets(insets);
    }];
    
    [self.indicateView mas_makeConstraints:^(MASConstraintMaker *make) {
        make.center.equalTo(self.pickerContainerView);
        make.height.mas_equalTo([JFFliterAppearanceManager sharedManager].pickerIndicaterSize.height);
        make.width.mas_equalTo([JFFliterAppearanceManager sharedManager].pickerIndicaterSize.width);
    }];
    
    [self.rightTF mas_makeConstraints:^(MASConstraintMaker *make) {
        make.top.equalTo(self.pickerContainerView).offset([JFFliterAppearanceManager sharedManager].pickerInsets.top);
        make.bottom.equalTo(self.pickerContainerView).offset([JFFliterAppearanceManager sharedManager].pickerInsets.bottom);
        make.right.equalTo(self.pickerContainerView).offset([JFFliterAppearanceManager sharedManager].pickerInsets.right);
        make.left.equalTo(self.indicateView.mas_right).offset(6);
    }];
    
    if (self.cellItem.sectionSelect) {
        [self.leftTF mas_remakeConstraints:^(MASConstraintMaker *make) {
            make.top.bottom.equalTo(self.rightTF);
            make.left.equalTo(self.pickerContainerView).offset([JFFliterAppearanceManager sharedManager].pickerInsets.left);
            make.right.equalTo(self.indicateView.mas_left).offset(-6);
        }];
    }else {
        [self.leftTF mas_remakeConstraints:^(MASConstraintMaker *make) {
            make.left.equalTo(self.pickerContainerView).offset([JFFliterAppearanceManager sharedManager].pickerInsets.left);
            make.top.bottom.right.equalTo(self.rightTF);
        }];
    }
}

#pragma mark - UITextFieldDelegate

- (BOOL)textFieldShouldBeginEditing:(UITextField *)textField {
    return YES;
}

- (BOOL)textField:(UITextField *)textField shouldChangeCharactersInRange:(NSRange)range replacementString:(NSString *)string {
    if (self.cellItem.pickerType != JFFliterPickerTypeTextField) {
        return NO;
    }
    NSString * toBeString = [textField.text stringByReplacingCharactersInRange:range withString:string];
    
    if ([toBeString isBlank]) {
        return YES;
    }
    
    if (![JFVarificationInputManager validateContentIsFloatNum:toBeString]) {
        return NO;
    }
    
    NSNumber *maxValue;
    if (textField == self.leftTF) {
        maxValue = self.cellItem.leftMaxNumber;
    }else {
        maxValue = self.cellItem.rightMaxNumber;
    }
    if (toBeString.doubleValue > maxValue.doubleValue || toBeString.length > ([NSString stringWithFormat:@"%d",(int)toBeString.doubleValue].length + 3)) {
        return NO;
    }
    return YES;
}

- (void)textFieldDidEndEditing:(UITextField *)textField {
    NSString *textValue = [textField.text isBlank] ? nil : textField.text;
    if (textField == self.leftTF) {
        self.cellItem.leftValue = textValue;
    }else {
        self.cellItem.rightValue = textValue;
    }
}


#pragma mark - Private

- (void)workWhenRunloopIdle:(void(^)())block {
    CFRunLoopRef runLoop = CFRunLoopGetCurrent();
    CFStringRef runLoopMode = kCFRunLoopDefaultMode;
    CFRunLoopObserverRef observer = CFRunLoopObserverCreateWithHandler(kCFAllocatorDefault, kCFRunLoopBeforeWaiting, true, 0, ^(CFRunLoopObserverRef observer, CFRunLoopActivity _) {
        if (block) {
            block();
        }
        CFRunLoopRemoveObserver(runLoop, observer, runLoopMode);
        CFRelease(observer);
    });
    CFRunLoopAddObserver(runLoop, observer, runLoopMode);
}

- (void (^)(NSString* str,NSDate * date))pickerBlockForTextField:(UITextField *)tf {
    @weakify(tf);
    return ^(NSString* str,NSDate * date){
        @strongify(tf);
        if (str&&date) {
            tf.text = str;
        }
        [tf resignFirstResponder];
    };
}

- (void (^)(NSString* province,NSString * city))addressPickerBlockForTextField:(UITextField *)tf {
    @weakify(tf);
    return ^(NSString* province,NSString * city){
        @strongify(tf);
        if (province&&city) {
            tf.text = [NSString stringWithFormat:@"%@,%@",province,city];
        }
        [tf resignFirstResponder];
    };
}

#pragma mark - Getter && Setter
- (UIView *)pickerContainerView {
    if (!_pickerContainerView) {
        _pickerContainerView = [[UIView alloc] init];
        _pickerContainerView.backgroundColor = [JFFliterAppearanceManager sharedManager].pickerContainerColor;
    }
    return _pickerContainerView;
}

- (UIView *)indicateView {
    if (!_indicateView) {
        _indicateView = [[UIView alloc] init];
        _indicateView.backgroundColor = [JFFliterAppearanceManager sharedManager].pickerIndicaterColor;
    }
    return _indicateView;
}

- (UITextField *)leftTF {
    if (!_leftTF) {
        _leftTF = [[UITextField alloc] init];
        _leftTF.textColor = [JFFliterAppearanceManager sharedManager].pickerTextColor;
        _leftTF.backgroundColor = [JFFliterAppearanceManager sharedManager].pickerBackgroundColor;
        _leftTF.borderStyle = UITextBorderStyleNone;
        _leftTF.textAlignment = NSTextAlignmentCenter;
        _leftTF.font = [JFFliterAppearanceManager sharedManager].pickerTextFont;
        _leftTF.delegate = self;
        _leftTF.keyboardType = UIKeyboardTypeDecimalPad;
    }
    return _leftTF;
}

- (UITextField *)rightTF {
    if (!_rightTF) {
        _rightTF = [[UITextField alloc] init];
        _rightTF.textColor = [JFFliterAppearanceManager sharedManager].pickerTextColor;
        _rightTF.backgroundColor = [JFFliterAppearanceManager sharedManager].pickerBackgroundColor;
        _rightTF.borderStyle = UITextBorderStyleNone;
        _rightTF.textAlignment = NSTextAlignmentCenter;
        _rightTF.font = [JFFliterAppearanceManager sharedManager].pickerTextFont;
        _rightTF.delegate = self;
        _rightTF.keyboardType = UIKeyboardTypeDecimalPad;
    }
    return _rightTF;
}

- (JFSelectDatePickView *)leftPickerView {
    if (!_leftPickerView) {
        _leftPickerView = [JFSelectDatePickView instanceView];
    }
    return _leftPickerView;
}

- (JFSelectDatePickView *)rightPickerView {
    if (!_rightPickerView) {
        _rightPickerView = [JFSelectDatePickView instanceView];
    }
    return _rightPickerView;
}

- (JFAreaPickerView *)leftAddressPicker {
    if (!_leftAddressPicker) {
        _leftAddressPicker = [[JFAreaPickerView alloc] init];
    }
    return _leftAddressPicker;
}

- (JFAreaPickerView *)rightAddressPicker {
    if (!_rightAddressPicker) {
        _rightAddressPicker = [[JFAreaPickerView alloc] init];
    }
    return _rightAddressPicker;
}

@end

File: GGSwift/JFFliter/Components/JFFliterSectionPickerCell.swift
//
//  JFFliterSectionPickerCell.swift
//  GGSwift
//
//  Created by geeksprite on 2017/8/1.
//  Copyright © 2017年 聚风天下有限公司. All rights reserved.
//

import UIKit

class JFFliterSectionPickerCell: JFFliterSectionCell {

    // MARK: Properties
    fileprivate lazy var leftTF: UITextField = {
        let tf = UITextField()
        tf.textColor = JFFliterAppearceManager.shared.pickerTextColor
        tf.backgroundColor = JFFliterAppearceManager.shared.pickerBackgroundColor
        tf.borderStyle = .none
        tf.textAlignment = .center
        tf.font = JFFliterAppearceManager.shared.pickerTextFont
        tf.delegate = self
        tf.keyboardType = .decimalPad
        return tf
    }()
    
    fileprivate lazy var rightTF: UITextField = {
        let tf = UITextField()
        tf.textColor = JFFliterAppearceManager.shared.pickerTextColor
        tf.backgroundColor = JFFliterAppearceManager.shared.pickerBackgroundColor
        tf.borderStyle = .none
        tf.textAlignment = .center
        tf.font = JFFliterAppearceManager.shared.pickerTextFont
        tf.delegate = self
        tf.keyboardType = .decimalPad
        return tf
    }()
    
    fileprivate lazy var pickerContainerView: UIView = {
        let view = UIView()
        view.backgroundColor = JFFliterAppearceManager.shared.pickerContainerColor
        return view
    }()
    
    fileprivate lazy var indicateView: UIView = {
        let view = UIView()
        view.backgroundColor = JFFliterAppearceManager.shared.pickerIndicaterColor
        return view
    }()
    
    fileprivate lazy var leftDateInputView: JFSelectDatePickView = {
        let picker = JFSelectDatePickView.instanceView()
        return picker
    }()
    
    fileprivate lazy var rightDateInputView: JFSelectDatePickView = {
        let picker = JFSelectDatePickView.instanceView()
        return picker
    }()
    
    fileprivate lazy var leftAddressInputView: JFSelectAddressPickerView = {
        return JFSelectAddressPickerView()
    }()
    
    fileprivate lazy var rightAddressInputView: JFSelectAddressPickerView = {
        return JFSelectAddressPickerView()
    }()
    
    // MARK: Override Methods
    override var cellItem: JFFliterItem? {
        didSet{
            guard cellItem != nil else {
                return
            }
            switch cellItem!.pickerType {
            case .DatePicker:
                self.wordWhenRunloopIdel {                    
                    self.leftDateInputView.myPickerView.minimumDate = self.cellItem?.leftMinDate
                    self.leftDateInputView.myPickerView.maximumDate = self.cellItem?.leftMaxDate
                    self.rightDateInputView.myPickerView.minimumDate = self.cellItem?.rightMinDate
                    self.rightDateInputView.myPickerView.maximumDate = self.cellItem?.rightMaxDate
                    self.leftTF.inputView = self.leftDateInputView
                    self.rightTF.inputView = self.rightDateInputView
                    self.leftDateInputView.callBack = self.datePickerBlock(textField: self.leftTF)
                    self.rightDateInputView.callBack = self.datePickerBlock(textField: self.rightTF)
                }
            case .AddressPicker:
                self.wordWhenRunloopIdel {
                    self.leftTF.inputView = self.leftAddressInputView
                    self.rightTF.inputView = self.rightAddressInputView
                    self.leftAddressInputView.callBack = self.addressPickerBlock(textField: self.leftTF)
                    self.rightAddressInputView.callBack = self.addressPickerBlock(textField: self.rightTF)
                }
                
            default:
                break
            }
            
            if let reset = self.cellItem?.shouldReset, reset {
                self.leftTF.text = nil
                self.rightTF.text = nil
            }
        }
    }
    
    override func didInit() {
        super.didInit()
        self.addSubview(self.pickerContainerView)
        self.pickerContainerView.addSubview(self.leftTF)
        self.pickerContainerView.addSubview(self.indicateView)
        self.pickerContainerView.addSubview(self.rightTF)
    }
    
    override func didLayoutSubviews() {
        super.didLayoutSubviews()
        self.pickerContainerView.frame = CGRect.init(x: JFFliterAppearceManager.shared.sectionPaddingSize.width, y: 0, width: self.JFFliter_width - 2.0 * (JFFliterAppearceManager.shared.sectionPaddingSize.width), height: self.JFFliter_height - JFFliterAppearceManager.shared.sectionPaddingSize.height)
    
        self.indicateView.frame = CGRect.init(origin: CGPoint.init(x: (self.pickerContainerView.JFFliter_width - JFFliterAppearceManager.shared.pickerIndicaterSize.width)/2.0 , y: (self.pickerContainerView.JFFliter_height - JFFliterAppearceManager.shared.pickerIndicaterSize.height)/2.0), size: (self.cellItem!.sectionSelect && self.cellItem!.pickerType != .AddressPicker) ? JFFliterAppearceManager.shared.pickerIndicaterSize : CGSize.zero)
        
        self.leftTF.frame = CGRect.init(x: JFFliterAppearceManager.shared.pickerInsets.left, y: JFFliterAppearceManager.shared.pickerInsets.top, width: (self.cellItem!.sectionSelect && self.cellItem!.pickerType != .AddressPicker) ? (self.indicateView.frame.minX - 6.0 - JFFliterAppearceManager.shared.pickerInsets.left) : (self.pickerContainerView.JFFliter_width - JFFliterAppearceManager.shared.pickerInsets.left + JFFliterAppearceManager.shared.pickerInsets.right), height: self.pickerContainerView.JFFliter_height - JFFliterAppearceManager.shared.pickerInsets.top + JFFliterAppearceManager.shared.pickerInsets.bottom)
        
        self.rightTF.frame = CGRect.init(x: self.indicateView.frame.maxX + 6, y: self.leftTF.JFFliter_y, width: (self.cellItem!.sectionSelect && self.cellItem!.pickerType != .AddressPicker) ? self.leftTF.JFFliter_width : 0 , height: self.leftTF.JFFliter_height)
    }
    
    // MARK: Private Methods
    private func wordWhenRunloopIdel(work:@escaping ()->Void) {
        let runloop = CFRunLoopGetCurrent()
        let runloopMode = CFRunLoopMode.defaultMode
        let observer = CFRunLoopObserverCreateWithHandler(kCFAllocatorDefault, CFRunLoopActivity.beforeWaiting.rawValue, true, 0) { (observer, activity) in
            work()
            CFRunLoopRemoveObserver(runloop, observer, runloopMode)
        }
        CFRunLoopAddObserver(runloop, observer, runloopMode)
    }
    
    private func datePickerBlock(textField: UITextField) -> (String?,Date?)->Void {
        func closure(str: String?, date: Date?) {
            if str != nil && date != nil {
                textField.text = str!
            }
            textField.resignFirstResponder()
        }
        return closure
    }
    
    private func addressPickerBlock(textField: UITextField) -> (String?,String?)->Void {
        func closure(str0: String?, str1: String?) {
            if str0 != nil && str1 != nil {
                textField.text = "\(str0!)  \(str1!)"
            }
            textField.resignFirstResponder()
        }
        return closure
    }
    
}

extension JFFliterSectionPickerCell : UITextFieldDelegate {

    func textField(_ textField: UITextField, shouldChangeCharactersIn range: NSRange, replacementString string: String) -> Bool {
        guard let pickerType = self.cellItem?.pickerType, pickerType == .TextField else {
            return false
        }

        if let toBeString = textField.text?.replacingCharacters(in: (textField.text?.JFFliter_range(from: range))!, with: string) {
            if toBeString.isEmpty {
                return true
            }
        
            if let doubleValue = Double(toBeString) {
                if let maxValue = textField.isEqual(self.leftTF) ? self.cellItem!.leftMaxNumber : self.cellItem!.rightMaxNumber {
                    if maxValue.doubleValue.isLess(than: doubleValue) || toBeString.characters.count > (String.init(format: "%d", Int(doubleValue)).characters.count + 3) {
                        return false
                    }
                }
            }else {
                return false
            }
            
            return true
        }

        return true
    }
    
    func textFieldDidEndEditing(_ textField: UITextField) {
        if textField.isEqual(self.leftTF) {
            self.cellItem!.leftValue = textField.text
        }else {
            self.cellItem?.rightValue = textField.text
        }
    }
}

//MARK: NSRange转化为range
extension String {
    func JFFliter_range(from nsRange: NSRange) -> Range<String.Index>? {
        guard
            let from16 = utf16.index(utf16.startIndex, offsetBy: nsRange.location, limitedBy: utf16.endIndex),
            let to16 = utf16.index(from16, offsetBy: nsRange.length, limitedBy: utf16.endIndex),
            let from = String.Index(from16, within: self),
            let to = String.Index(to16, within: self)
            else { return nil }
        return from ..< to
    }
}
